### PR TITLE
Mailer - support Qute templates

### LINKED ISSUE #6158: Support Qute templates in mailer extension
API example:

```java
@Inject
MailTemplate test;

CompletionStage<Void> send() {
   // Render the templates src/main/resources/templates/test.html (and test.txt) as the message body
   return test.to("quarkus@quarkus.io").subject("Test").data("name", "John").send();
}
```

### LINKED ISSUE #6159: Qute - collect all possible template variants
Currently, the template variants are only collected for `VariantTemplate` injection points. However, it should be also possible to obtain a `VariantTemplate` instance programmatically.

### PR DESCRIPTION
**Pull request by mkouba**
- resolves #6158
- also collect all possible template variants and fix #6159

### CODE REVIEW

**Review by cescoffier — APPROVED** [review]
Pretty cool! 

Should we update the quickstart?

**Review by FroMage — COMMENTED** [review]
Yeah, beautiful, spot on!

Needs docs, but otherwise great!

File: extensions/mailer/deployment/src/main/java/io/quarkus/mailer/deployment/MailerProcessor.java [hunk 1]
```diff
@@ -68,4 +87,50 @@ MailerBuildItem build(BuildProducer<FeatureBuildItem> feature, MailConfigRecorde
...
+        for (InjectionPointInfo injectionPoint : validationPhase.getContext().get(BuildExtension.Key.INJECTION_POINTS)) {
+            if (injectionPoint.getRequiredType().name().equals(MAIL_TEMPLATE)) {
+                AnnotationInstance resourcePath = injectionPoint.getRequiredQualifier(QuteProcessor.RESOURCE_PATH);
+                String name;
+                if (resourcePath != null) {
+                    name = resourcePath.value().asString();
+                } else if (injectionPoint.hasDefaultedQualifier()) {
+                    name = QuteProcessor.getName(injectionPoint);
+                } else {
+                    name = null;
```
Comment (gsmet) [inline extensions/mailer/deployment/src/main/java/io/quarkus/mailer/deployment/MailerProcessor.java]: This is a valid case?

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/MailTemplate.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+        MailTemplateInstance subject(String subject);
+
+        MailTemplateInstance from(String from);
+
+        MailTemplateInstance replyTo(String replyTo);
+
+        MailTemplateInstance bounceAddress(String bounceAddress);
+
+        MailTemplateInstance data(String key, Object value);
```
Comment (gsmet) [inline extensions/mailer/runtime/src/main/java/io/quarkus/mailer/MailTemplate.java]: Let's also plan to include the `Locale` at some point.

**Review by gsmet — COMMENTED** [review]
I added a small question. And a suggestion but I'm pretty sure you already have that one in mind.

**Comment by geoand** [comment]
:heart: 

I definitely agree we should have docs and a quickstart

**Comment by gsmet** [comment]
Maybe let's add something to the existing mailer quickstart instead of having a specific one? That would have my preference.

**Comment by geoand** [comment]
Good idea!

**Comment by geoand** [comment]
The Windows build seems to be failing (I already tried restarting it to see if it was a flake):

```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.447 s <<< FAILURE! - in io.quarkus.mailer.InjectionTest
[ERROR] testInjection  Time elapsed: 0.063 s  <<< ERROR!
java.lang.NullPointerException
	at io.quarkus.qute.runtime.VariantTemplateProducer$VariantTemplateInstanceImpl.<init>(VariantTemplateProducer.java:131)
	at io.quarkus.qute.runtime.VariantTemplateProducer$VariantTemplateImpl.instance(VariantTemplateProducer.java:105)
	at io.quarkus.mailer.runtime.MailTemplateProducer$2.instance(MailTemplateProducer.java:75)

```

**Comment by stuartwdouglas** [comment]
Please don't restart tests like this where the test has been modified as part of the PR. Best case it just wastes some CI resources and it fails again, but worst case is the test actually passes on the second run and we accidentally merge a flakey test.

**Comment by geoand** [comment]
Sorry @stuartwdouglas! I must admit that when I restarted the failing Windows I did so without looking at the logs to see what actually failed (and wrongfully assumed it was one of the flakes we've seen over time).
You are absolutely right, I'll be more careful from now on

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/MailTemplate.java (hunk 2 above)
Comment (mkouba) [reply extensions/mailer/runtime/src/main/java/io/quarkus/mailer/MailTemplate.java]: Yes, that's definitely on the plan... once we support `Locale` in templates in general ;-).

File: extensions/mailer/deployment/src/main/java/io/quarkus/mailer/deployment/MailerProcessor.java (hunk 1 above)
Comment (mkouba) [reply extensions/mailer/deployment/src/main/java/io/quarkus/mailer/deployment/MailerProcessor.java]: It is, in theory. We do not cover injection points with `MailTemplate` required type and qualifiers other than `@ResourcePath` or `@Default` (implicit).

**Comment by mkouba** [comment]
The PR should be ready now...

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
Thanks Martin. It's nice to see all this put in motion.

**Pull request merged**